lib/process_settings/abstract_monitor.rb

```diff
@@ -94,10 +96,31 @@ def static_context=(context)
     def targeted_value(*path, dynamic_context:, required: true)
       # Merging the static context in is necessary to make sure that the static context isn't shifting
       # this can be rather costly to do every time if the dynamic context is not changing
-      # TODO: Warn in the case where dynamic context was attempting to change a static value
-      # TODO: Cache the last used dynamic context as a potential optimization to avoid unnecessary deep merges
-      # TECH-4402 was created to address these todos
-      full_context = dynamic_context.deep_merge(static_context)
+
+      # Warn in the case where dynamic context was attempting to change a static value
+      changes = dynamic_context.each_with_object({}) do |(key, dynamic_value), result|
+        if static_context.has_key?(key)
+          static_value = static_context[key]
+          if static_value != dynamic_value
+            result[key] = [static_value, dynamic_value]
+          end
+        end
+      end
+
+      changes.empty? or warn("WARNING: static context overwritten by dynamic!\n#{changes.inspect}")
+
+      # Cache the last used dynamic context as a potential optimization to avoid unnecessary deep merge
+      if full_context_cache[dynamic_context]
+        full_context = full_context_cache[dynamic_context]
+        logger.info("cache hit with dynamic_context: #{dynamic_context}")
+      else
+        full_context = dynamic_context.deep_merge(static_context)
+        if full_context_cache.size <= 1000 # keep this cache memory finite
+          full_context_cache[dynamic_context] = full_context
+        end
+        logger.info("cache miss with dynamic_context: #{dynamic_context}")
+      end
```

Comment:
If you break this block of code out into it's own method with the sole responsibility of accessing the full context from the cache, this becomes a lot cleaner.

```ruby
  full_context = full_context_from_cache(dynamic_context)
  ...
end

def full_context_from_cache(dynamic_context)
  if (full_context = full_context_cache[dynamic_context])
    logger.info("cache hit ...")
    full_context
  else
    logger.info("cache miss ...")
    dynamic_context.deep_merge(static_context).tap do |full_context|
      if full_context_cache.size <= 1000
        full_context_cache[dynamic_context] = full_context
      else
        logger.info("cache limit reached ...")
      end
    end
  end
end
```